From 302532f86de2e154b9468061a6b4c3ff6925241a Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 11:59:06 -0400
Subject: [infiniband] cm: Fix missing unlock on error in cm_init_qp_rtr_attr()

Message-id: <2f93f5686175166d99c6522df52cb3093007d4ad.1410425016.git.avadai@redhat.com>
Patchwork-id: 93594
O-Subject: [Patch RHEL7.1 V1 010/197] IB/cm: Fix missing unlock on error in cm_init_qp_rtr_attr()
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Wei Yongjun <yongjun_wei@trendmicro.com.cn>

Add the missing unlock before return from function cm_init_qp_rtr_attr()
in the error handling case.

Fixes: dd5f03beb4f7 ("IB/core: Ethernet L2 attributes in verbs/cm structures")
Signed-off-by: Wei Yongjun <yongjun_wei@trendmicro.com.cn>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 990acea616e99355703b503c1e50fb9c7ddff6b9)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/core/cm.c b/drivers/infiniband/core/cm.c
index eb8de66..c323917 100644
--- a/drivers/infiniband/core/cm.c
+++ b/drivers/infiniband/core/cm.c
@@ -3509,8 +3509,10 @@ static int cm_init_qp_rtr_attr(struct cm_id_private *cm_id_priv,
   *qp_attr_mask = IB_QP_STATE | IB_QP_AV | IB_QP_PATH_MTU |
     IB_QP_DEST_QPN | IB_QP_RQ_PSN;
   qp_attr->ah_attr = cm_id_priv->av.ah_attr;
-  if (!cm_id_priv->av.valid)
+  if (!cm_id_priv->av.valid) {
+   spin_unlock_irqrestore(&cm_id_priv->lock, flags);
    return -EINVAL;
+  }
   if (cm_id_priv->av.ah_attr.vlan_id != 0xffff) {
    qp_attr->vlan_id = cm_id_priv->av.ah_attr.vlan_id;
    *qp_attr_mask |= IB_QP_VID;
-- 
1.7.1